CatLover assignment: Orestis Zinelis

OrestisZinelis opened the pull request:
# CatLover 

A React application for **cat lovers**. Display cat images, mark them as favorites and check the breed info!

## 📸 Screenshots

![CatLovers-Cats](https://github.com/user-attachments/assets/7d67802c-9900-4ac0-9806-29d445a17073)
![CatLovers-Breeds](https://github.com/user-attachments/assets/5cc21db4-f6c5-454e-9367-db013787e35b)
![CatLovers-Cats-Responsive](https://github.com/user-attachments/assets/8cb6c2d3-b81b-4164-a5de-945e26fab206)
![CatLovers-Breed-Modal](https://github.com/user-attachments/assets/0bfb8902-ab52-4032-a0d3-8fd2675d4313)
![CatLovers-Cat-Modal](https://github.com/user-attachments/assets/65137cdb-24c9-4f7e-8be5-c9976e7a1ec3)

## Tech Stack

- React
- React Router
- TypeScript
- Vite
- Axios
- TanStack Query
- Tailwind CSS
- Material UI

## API

Get your API key from [The Cat API](https://developers.thecatapi.com/), then set `VITE_CAT_API_KEY` inside the `.env` file.

## Installation

1. Clone the repo

```sh
git clone https://github.com/OrestisZinelis/cat-lover
```

2. Navigate to the project directory e.g.

```sh
cd cat-lover
```

3. Install dependencies

```sh
yarn
```

or

```sh
npm install
```

4. Create a `.env` file in the root directory and add your API key:

`VITE_CAT_API_KEY=your_api_key_here`

## Development

Start the development server:

```sh
yarn dev
```

This will launch the app on http://localhost:5173

## Build

Build the project for production:

```sh
yarn build
```

This will generate the build files in the dist/ directory.

## Linting

Lint the project files using ESLint:

```sh
yarn lint
```

## Preview

Preview the built project:

```sh
yarn preview
```

OrestisZinelis committed:
Init project

OrestisZinelis committed:
Add tailwind, react-query, axios, react-router

OrestisZinelis committed:
Add Material UI

OrestisZinelis committed:
Bump TypeScript version

OrestisZinelis committed:
Set logo and title

OrestisZinelis committed:
Cleanup Css

OrestisZinelis committed:
Set Paths

OrestisZinelis committed:
Add entities types

OrestisZinelis committed:
Add axios api handlers

OrestisZinelis committed:
Add entities queries

OrestisZinelis committed:
Add AppTopBar

OrestisZinelis committed:
Edit gitignore not include .env

OrestisZinelis committed:
Add Cat & Breed Modals

OrestisZinelis committed:
Add Breeds page

OrestisZinelis committed:
Add Cats page

OrestisZinelis committed:
Add Favorites page

OrestisZinelis committed:
Add SkeletomImages component

OrestisZinelis committed:
Add responsive image columns hook

OrestisZinelis committed:
Add App.tsx, main.tsx

OrestisZinelis committed:
Add README.md

OrestisZinelis committed:
Merge remote-tracking branch 'catlover/main' into feature

padelis commented on src/App.tsx:
```diff
@@ -0,0 +1,63 @@
+    () => (message: string, severity: 'success' | 'error') => {
+      setAlert({ open: true, message, severity })
+    },
+    [setAlert]
```
Is `setAlert` actually needed here?

OrestisZinelis replied:
You are right, it should not be included in the dependencies array. It will never change across renders, because it is from `useState`.

padelis commented on package.json:
```diff
@@ -0,0 +1,39 @@
+    "@emotion/styled": "^11.14.0",
+    "@fontsource/roboto": "^5.1.1",
+    "@mui/icons-material": "^6.4.2",
+    "@mui/material": "^6.4.2",
+    "@tailwindcss/vite": "^4.0.0",
```
How come you chose both `mui` and `tailwind` together?

OrestisZinelis replied:
I started the project using tailwind and then took the decision to use material UI 😅 . 

In some places I could use the sx props or theme override. I could have done a better job at this, be more consistent.

padelis reviewed (commented):
Hello @OrestisZinelis! 👋 
First of all, thanks for taking the time and effort to do the assignment! 

I've added a couple of questions if you wanna take a look. 
Also have you used any tools to format your code?

OrestisZinelis committed:
Add prettier, format everything

OrestisZinelis commented:
> Hello @OrestisZinelis! 👋 First of all, thanks for taking the time and effort to do the assignment!
> 
> I've added a couple of questions if you wanna take a look. Also have you used any tools to format your code?

I have used Prettier (via VSCode extension) but I did not install prettier and configure pretier with eslint...  I added a commit

OrestisZinelis committed:
Fix `showAlert`, remove `setAlert` dependency

padelis commented:
Thanks again for the prompt replies and fixes

gpositive commented on package.json:
```diff
@@ -0,0 +1,43 @@
+  "dependencies": {
+    "@emotion/react": "^11.14.0",
+    "@emotion/styled": "^11.14.0",
+    "@fontsource/roboto": "^5.1.1",
```
Is this font-family actually needed? I can't see this font being used anywhere but I might be wrong.

OrestisZinelis replied:
```
<style data-emotion="css-global" data-s="">body{margin:0;color:rgba(0, 0, 0, 0.87);font-family:"Roboto","Helvetica","Arial", sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;background-color:#fff;}</style>
```

Material UI applies it. I followed the instructions about the Roboto font from the Material UI documentation. The font is used but is not visible from the code source.

gpositive commented on src/pages/Cats.tsx:
```diff
@@ -0,0 +1,47 @@
+                <img src={cat.url} alt="Cat" loading="lazy" />
+                <ImageListItemBar
+                  position="bottom"
+                  title={cat?.breeds[0]?.name ?? ""}
```
There was a case where I got a white screen of death that was coming from this line.
Unfortunately I haven't managed to debug in depth.
Can you guess why this might have happened?

OrestisZinelis replied:
I cannot reproduce it, unfortunately.

In the api request I use `has_breeds=1`, I get only cats with available breed, but needs a valid API key. The main problem is that I do not handle any possible fetch error. I initialize the cats as an empty array, but the code will crash if `breeds` is not an array. 

I prepared a commit to handle fetching errors or empty data.

gpositive commented on src/components/BreedModal.tsx:
```diff
@@ -0,0 +1,108 @@
+                  color="secondary"
+                  onClick={() => refetchCats()}
+                >
+                  Load more
```
Correct me if I'm wrong but right now, clicking 'Load More' fetches the same data repeatedly. Is this intended behaviour? If not are there any ways to improve it?

OrestisZinelis replied:
This is a tricky one. It depends on the breed. If there are more than 10 images available for a breed, then the user may get different images on refetch (e.g. Maine Coon breed).

gpositive commented on src/api/favorite.api.ts:
```diff
@@ -0,0 +1,27 @@
+    image_id: imageId,
+  });
+
+  const cat = await getCatById(imageId); // Fetch the image URL associated with this imageId to avoid refetching favorites
```
Could you please explain this a bit more?

OrestisZinelis replied:
Each time I add a favorite, I also fetch the cat to update the imageUrl in `useAddFavoriteMutation`.

```ts
 return oldFavorites.map((fav) =>
            fav.image_id === imageId
              ? {
                  ...fav,
                  id: data.id,
                  image_id: imageId,
                  image: { id: imageId, url: data.imageUrl },
                }
              : fav,
          );
```
          
This way, the imageUrl will exist for an added favorite and the favorites data will be consistent, without the need to fetch favorites (update oldFavorites array optimistically). 

Pros: favorites data is not refetched each time we access the favorites page
Cons: An additional get request each time user adds a favorite.

gpositive commented on src/components/BreedModal.tsx:
```diff
@@ -0,0 +1,108 @@
+                  (cats || []).map((cat) => (
+                    <ImageListItem key={cat.id}>
+                      <Link to={`./cats/${cat.id}`} relative="path">
+                        <img src={cat.url} alt="Cat" loading="lazy" />
```
Can you explain what loading="lazy" does and why you chose to use it here?

OrestisZinelis replied:
`lazy` will delay the loading of the image until it is about to scroll into view. I should not have used it here. If we fetch many images, that are not visible to the user then `lazy` should be used.

gpositive commented on src/hooks/queries/cat.queries.use.ts:
```diff
@@ -0,0 +1,26 @@
+  useQuery({
+    queryKey: ["cats-by-breed", breedId],
+    queryFn: () => getCatsByBreedId(breedId, 10),
+    refetchOnMount: false,
```
In some of these `useQuery` params you're setting `refetchOnMount: false` but not in others. What are the reasons behind this decision?

OrestisZinelis replied:
I use everywhere `refetchOnMount: false` except  in `useGetCatById` query which is a mistake (I missed it)

For Cats page: I did not want to refetch random cats, only onClick load more button.
For Breeds page: (table) I did not want to refetch data. Breeds data won't change so often.
For Favorites page: I update oldFavorites array optimistically, there is no need to fetch favorites on mount

I pushed a fix  for `useGetCatById`

gpositive commented on src/layouts/AppTopBar.tsx:
```diff
@@ -0,0 +1,153 @@
+              display: { xs: "none", md: "flex" },
+              fontFamily: "monospace",
+              fontWeight: 700,
+              letterSpacing: ".3rem",
```
What's the difference between `rem` and `em`? Why you've chosen to go with `rem` in this specific example?

OrestisZinelis replied:
To implement the top bar, I used [this](https://mui.com/material-ui/react-app-bar/?srsltid=AfmBOop09MlaruqQPVysIf0fJA2XDuvERMlk7Xw0-GGGFi1zVUnUodpx#system-ResponsiveAppBar.tsx) example. I did not check the code thoroughly. 

In general, the root element (rem) is relative to the root <HTML>. em is relative to the font size of the parent element.

I guess that for a fixed, consistent look like a logo or branding element, rem is the right choice.

gpositive reviewed (commented):
Heya Oresti! 👋

Thank you for taking the time to complete this assignment! We really appreciate it!

I've already added some comments and some more questions for you to answer if you can.

- Is there anything you would improve in this assignment if you had more time?
- Testing is missing from your project. If you were to test your application, how would you approach this?

Thanks again!

OrestisZinelis committed:
Handle errors in pages

OrestisZinelis committed:
Add AlertContext

OrestisZinelis committed:
Declare content types

OrestisZinelis committed:
Use direct paths (not @src)

OrestisZinelis committed:
Fix useGetCatById `refetchOnMount: false`

OrestisZinelis committed:
Remove `lazy` in `BreedModal`

OrestisZinelis commented:
> Heya Oresti! 👋
> 
> Thank you for taking the time to complete this assignment! We really appreciate it!
> 
> I've already added some comments and some more questions for you to answer if you can.
> 
> * Is there anything you would improve in this assignment if you had more time?
> * Testing is missing from your project. If you were to test your application, how would you approach this?
> 
> Thanks again!

 I would focus on improving the user experience, accessibility, and performance of the application. 

For testing, mocking API responses and testing edge cases would also be a priority to cover all possible scenarios. I would use a combination of unit, integration, and E2E tests.


**Responsive Design:** Now the app is somewhat responsive for 600px min width.
**Accessibility:** e.g. Add keyboard navigation
**Breeds table:** Search & filter & pagination for breeds
**Better favorites management:** Display some statistics, group by breed, toggle between list/grid view modes, batch deletion
Add **Localization** 
**User Authentication**
